programs/clearing_house/src/lib.rs

```diff
@@ -754,6 +757,226 @@ pub mod clearing_house {
+        market.amm.net_base_asset_amount = market
+            .amm
+            .net_base_asset_amount
+            .checked_add(lp_position.base_asset_amount)
+            .ok_or_else(math_error!())?;
+
+        let base_asset_acquired = lp_position.base_asset_amount;
+        let quote_asset_amount = lp_position.quote_asset_amount;
+
+        if lp_position.base_asset_amount > 0 {
+            market.base_asset_amount_long = market
+                .base_asset_amount_long
+                .checked_add(base_asset_acquired)
+                .ok_or_else(math_error!())?;
+            market.amm.quote_asset_amount_long = market
+                .amm
+                .quote_asset_amount_long
+                .checked_add(quote_asset_amount)
+                .ok_or_else(math_error!())?;
+        } else {
+            market.base_asset_amount_short = market
+                .base_asset_amount_short
+                .checked_add(base_asset_acquired)
+                .ok_or_else(math_error!())?;
+            market.amm.quote_asset_amount_short = market
+                .amm
+                .quote_asset_amount_short
+                .checked_add(quote_asset_amount)
+                .ok_or_else(math_error!())?;
+        }
+
+        market.open_interest = market
+            .open_interest
+            .checked_add(1)
+            .ok_or_else(math_error!())?;
+
+        // update market state
+        let new_sqrt_k = market
+            .amm
+            .sqrt_k
+            .checked_sub(lp_tokens_to_burn)
+            .ok_or_else(math_error!())?;
+        let new_sqrt_k_u192 = bn::U192::from(new_sqrt_k);
+
+        let update_k_result = get_update_k_result(&market, new_sqrt_k_u192, false)?;
+        math::amm::update_k(&mut market, &update_k_result)?;
+
+        Ok(())
+    }
+
+    #[access_control(
+        exchange_not_paused(&ctx.accounts.state)
+    )]
+    pub fn add_liquidity<'info>(
+        ctx: Context<AddRemoveLiquidity>,
+        quote_asset_amount: u128,
+        market_index: u64,
+    ) -> Result<()> {
+        let user = &mut load_mut(&ctx.accounts.user)?;
+        let clock = Clock::get()?;
+        let remaining_accounts_iter = &mut ctx.remaining_accounts.iter().peekable();
+
+        let mut oracle_map = OracleMap::load(remaining_accounts_iter, clock.slot)?;
+        let bank_map = BankMap::load(
+            &get_writable_banks(QUOTE_ASSET_BANK_INDEX),
+            remaining_accounts_iter,
+        )?;
+        let market_map = MarketMap::load(
+            &get_writable_markets(market_index),
+            &MarketOracles::new(), // TODO
+            remaining_accounts_iter,
+        )?;
+
+        let position_index = get_position_index(&user.positions, market_index)
+            .or_else(|_| add_new_position(&mut user.positions, market_index))?;
+        let lp_position = &mut user.positions[position_index];
+
+        // TODO: allow lps to add to position when settle is working
+        let can_lp = !(lp_position.has_open_order()
```

Comment:
nitpick: can we add a `can_lp` function to `MarketPosition` struct. This way, if another dev changes `has_open_order` `is_open_position` or `is_lp`, they're more likely to realize they need to account for can_lp

Reply:
can_lp is just tmp since now we are letting lps trade you should always be able to lp - next iter it should be gone